---
title: "Chapter 1: Cell Type Annotation of Integrated Control Samples"
author: "Joanna Xie"
date: "`r doc_date()`"
bibliography: ../data/chap1_bib.bib
output:
  BiocStyle::html_document:
    toc: yes
    toc_depth: 3
    toc_float: yes
    code_folding: "hide"
citation_package: natbib  
csl: ../data/nature.csl  
---


```{r setup, include=FALSE}

knitr::opts_chunk$set(
  tidy = TRUE,
  tidy.opts = list(width.cutoff = 95),
  fig.width = 15,
  fig.height = 11,
  warning = FALSE,
  message = FALSE,
  verbose = FALSE,
  root.dir = "../html")
 
 
set.seed(1996)
```

```{r Libraries, echo=FALSE}

library(dplyr)
library(ggplot2)
library(tidyverse)
library(Seurat)
library(cowplot)
library(patchwork)
library(scDblFinder)
library(speckle)

```


```{r defineAesthetics, include=FALSE, warning = FALSE }

source("../vignettes/themes_colors.R")

#sample colours 
#Wes Anderson
sampleColors <- c(
#Color 1: "Vintage Yellow"
"1. HBEC D1 Control"  = "#f0a60c", #(Light)

#Color 2: "Soft Salmon"
"4. COPD D1 Control"  = "#FF85AB", #(Light)

#Color 3: "Dusty Lavender"
"7. COPD D2 Control"  = "#09717d") #(Light)

#Cell Colours
cellColors <- c(
  "Ciliated" = "#e69317",        
  "Club" = "#56c0f5",           
  "Basal" = "#059971",          
  "Basal-Club Transitional" = "#f0e22b",        
  "Ciliated (Stressed)" = "#0540b5", 
  "Proliferating Basal" = "#8a491a", 
  "Proliferating + Ciliated" = "#c2559c", 
  "Ionocyte + Tuft" = "#7bd938",        
  "Goblet" = "#535366"
)

```


# Summary


This vignette models the cell types of human bronchial lung epithelium in vitro using air-liquid interface (ALI) cultures of hBEC cells derived from patients. Annotation of scRNA-seq data demonstrates that the major cell types found in vivo (e.g., goblet, ciliated) are replicated in vitro within the pseudostratified epithelial structure, based on the expression of canonical biomarkers.

The following steps were taken:

* Dataset was loaded. Dataset consists of three scRNAseq libraries, one for each hBEC ALI culture from three different patients. Of these patients, two are diagnosed with COPD.
* Quality of each dataset was assessed and a set of consistent quality control metrics were applied to each library accordingly.
* Each library was independently normalised and variable features were selected.
* Anchors were used to create an integrated data object.
* The integrated object was analysed with PCA and UMAP.
* Clustering was used to identify cell types.
* Manual annotation was performed by analysing expression levels of known biomarkers of human bronchial epithelium cell types.
* Further biomarkers were obtained computationally by Seurat analyses.
* Additional plots: feature plots and bar plot of cell-type proportions.

The analyses clearly demonstrate that the ALI cultures developed cell types resembling those found in vivo in bronchial epithelium, to the extent that clusters could be manually annotated based on scRNA-seq data. However, cluster annotation can be ambiguous, as some clusters express markers of multiple cell types, and stressed or dying cell populations may persist even after quality control.

A continuum of biomarker expression was observed—for example, cells annotated as club cells express goblet cell biomarkers, and there is overlap between proliferating basal cells and ciliated cells. As a result, quantitative comparisons between cell types require careful consideration.

# Loading Data

Data was loaded, and metadata including sample type and treatment added.

```{r loadData}

metafile<- read.csv("../data/meta_data.csv")

metafile[c("Number", "disease", "D", "Treatment")] <- str_split_fixed(metafile$Sample.name, ' ', 4)

metafile <- filter(metafile, Barcode %in% c("SITTH1","SITTC2","SITTF2"))

# List names of all  libraries 
libs = c("SITTH1",
         "SITTC2",
         "SITTF2") 

# For each library name, get file path to read in the 10x data, then create Seurat Object.
ctrl_list <- lapply(libs, function(x){
  
# File path
filename <- paste0("/data/jx303/processedData_2024ref/", x , "/outs/filtered_feature_bc_matrix/")
  
# 10x data
data10x <- Read10X(data.dir = filename)
  
# Creating Seurat Object 
obj <- CreateSeuratObject(counts = data10x, project = x, min.cells = 3, min.features = 100)
 
# Extracts the orig.ident column as dataframe, renames the single column in the df as "Barcode"
cell_sampleIDs <- as.data.frame(obj$orig.ident)
colnames(cell_sampleIDs) <- "Barcode"
cellIDs <- row.names(cell_sampleIDs) # Saves cellIDs variable from the row names of cell_samples_IDs

#Add meta data to cell IDs for each library
metaData <- left_join(cell_sampleIDs, 
                      #add the data for this specific sampleID
                      filter(metafile, Barcode ==x), by = "Barcode") 
row.names(metaData) <- cellIDs
# metaData has as many rows as there are cells in the libraries. The contents of each row for a given library are identical (sample metadata), only the row names, which gives the cell IDs, differ. 

#add three meta data columns as three separate slots to each Seurat object 
metaData$disease <- as.factor(metaData$disease)
obj$tissueType <- metaData[6]

metaData$Sample.name <- as.factor(metaData$Sample.name)
obj$sampleType <- metaData[4]

metaData$Treatment <- as.factor(metaData$Treatment)
obj$treatment <- metaData[8]

DefaultAssay(obj) <- "RNA"

obj
})

names(ctrl_list) <- libs
```


# Quality Control

## Visualising Data Quality


* A high proportion of mitochondrial RNA (mtRNA) is often indicative of a stressed or dying cell, hence we ignore these cells for downstream analysis. The set of genes beginning MT- is used to define the mitochondrial genes.
* An abnormally high number of (unique gene) features suggests doublets or multiplets. Conversely, an abnormally low feature count suggests low cell quality or empty droplet. 
* The gene feature counts correlate strongly with total count number, as shown in the feature scatter plots, so this metric may be used for QC filtering in a similar manner as feature count. The number in the title shows the Pearson correlation coefficient.

```{r mt, fig.width = 12, fig.height = 8}


# In the Cell Ranger output, identified mitochondrial genes mostly start with MT-, which we use as the pattern.

ctrl_list <-  lapply(ctrl_list, function(obj){
  obj[["percent.mt"]] <- PercentageFeatureSet(obj, pattern = "^MT-")
  obj
})

# Violin plots to estimate suitable cut-offs for QC filtering.

lapply(ctrl_list, function(x){
  p <- VlnPlot(x, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
  p[[1]] <- p[[1]] + ggtitle("Number of Features") + ylab("Number of unique features")
  p[[2]] <- p[[2]] + ggtitle("Number of Counts") + ylab("Number of counts")
  p[[3]] <- p[[3]] + ggtitle("% Mitochondrial Count") + ylab("Percentage of mitochondrial reads")
  sample_type <- as.character(unique(x$sampleType))
  p <- p + plot_annotation(title = paste0("Violin plots for QC metrics: ", sample_type))
  sample_type <- as.character(unique(x$sampleType))
  p

})

lapply(ctrl_list, function(x){
  plot1 <- FeatureScatter(x, feature1 = "nCount_RNA", feature2 = "percent.mt") + ylab("Percentage of mitochondrial reads") + xlab("Number of counts")
  plot2 <- FeatureScatter(x, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")  + ylab("Number of unique features") + xlab("Number of counts")
  sample_type <- as.character(unique(x$sampleType))
  plot1 + plot2 + plot_annotation(title = paste0("Feature scatter plots for QC metrics: ", sample_type))
})

```

## Filtering

After manual inspection of the data, the following QC metrics were used to subset the data:

* Removed cells with more than 15% mitochondrial RNA contamination. This is a high threshold, but is data-driven, suggesting an elevated number of stressed/dying cells in the culture.
* Removed cells with less than 1000 unique feature counts.
* Removed cells with total counts either less than 5000 or more than 100,000.


```{r}

ctrl_list_QC <- lapply(ctrl_list, function(x){
  subset(x, subset = nFeature_RNA > 1000 & nCount_RNA < 100000 & nCount_RNA > 5000 & percent.mt < 15)
})
  
```

## Doublet Removal

Using default settings for scDblFinder, the following percentages of doublets were found for the libraries: 

* HBEC D1 Control - 5.5 %
* COPD D1 Control - 9.8 %
* COPD D2 Control - 8.3 %


```{r}

ctrl_list_QC <- lapply(ctrl_list_QC, function(obj) {
  sce <- as.SingleCellExperiment(obj)
  
  # Run scDblFinder to detect doublets
  sce <- scDblFinder(sce)
  
  # Add doublet classification to the Seurat metadata
  obj$doublet_class <- sce$scDblFinder.class
  
  # Filter out doublets
  obj <- subset(obj, subset = doublet_class == "singlet")
})

```

# Normalisation and Variable Features

Data was scaled and normalised with the default method (scaled by expression, multiplied by scale factor of 10,000, then log-transformed). Then variable features were found with vst method and the top 2000 variable features across libraries selected for integration. 


```{r}
ctrl_list_filtered <- lapply(ctrl_list_QC, function(x) {
  
  x <- NormalizeData(x)
  x <- FindVariableFeatures(x, selection.method = "vst", nfeatures = 2000)

  })

# Top 2000 repeatedly variable features across libraries selected for integration
features <- SelectIntegrationFeatures(object.list = ctrl_list_filtered)

```

## Variable Feature Plots

```{r}
lapply(1:length(ctrl_list_filtered), function(i) {

   x <- ctrl_list_filtered[[i]]
  
  top10 <- head(VariableFeatures(x), 10)

  plot1 <- VariableFeaturePlot(x, pt.size=0.7)
  plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
  sample_type <- as.character(unique(x$sampleType))
  plot1 + plot2 + plot_annotation(title = names(ctrl_list_filtered[i])) + plot_annotation(title=paste0("Variable features of: ", sample_type))
  
})
```


# Integration

Seurat was used to identify cellular anchors for integration.

```{r, include = FALSE}

# Attempting to troubleshoot warnings in FindIntegrationAnchors steps - not sure why layer counts are not present given that counts are present in the default assay that has been set as RNA. 

DefaultAssay(ctrl_list_filtered[[1]])
DefaultAssay(ctrl_list_filtered[[2]])
DefaultAssay(ctrl_list_filtered[[3]])


seurat_object <- ctrl_list_filtered[[3]]
# Check if counts slot exists in the RNA assay
if (is.null(seurat_object@assays$RNA@counts)) {
  cat("Counts layer is missing from the RNA assay")
} else {
  cat("Counts layer is present in the RNA assay")
}

```


```{r}

if(file.exists("../data/ctrl_filtered_anchors.RData")){
  
  load(file = "../data/ctrl_filtered_anchors.RData")

  }else{
    anchorsControl <- FindIntegrationAnchors(object.list = ctrl_list_filtered, anchor.features = features)
    
    save(anchorsControl, file = "../data/ctrl_filtered_anchors.RData")

  }


if(file.exists("../data/ctrl_integrated.RData")){
  
  load(file = "../data/ctrl_integrated.RData")
  
  }else{
    
    controllibsCombined <- IntegrateData(anchorset = anchorsControl)
    
    save(controllibsCombined, file = "../data/ctrl_integrated.RData")

}
```
# Dimensionality Reduction

## Scaling

Data was scaled and variation due to mitochondrial content regressed out. 

```{r}
DefaultAssay(ctrl_integrated) <- "integrated"

ctrl_integrated <- ScaleData(ctrl_integrated, verbose = FALSE, vars.to.regress = 'percent.mt')

```
## PCA

```{r,fig.width=10, fig.height=8}
ctrl_integrated <- RunPCA(ctrl_integrated, npcs = 35, verbose = FALSE)

DimPlot(ctrl_integrated, reduction = "pca", cols = NULL,  group.by = "sampleType") + 
  theme(legend.position = "right")  + 
  scale_color_manual(values = sampleColors) + ggtitle("PC1 and PC2 plot of integrated control samples")
```

### Heatmap 

Heatmaps for the top 6 PCs.
Each column is a cell and each row is a gene (the genes with the highest loadings for that PCA are displayed). The colour indicates the expression level of that gene for a given cell (higher expression going from magenta to black to yellow).

```{r,fig.width=16, fig.height=8}
DimHeatmap(ctrl_integrated, dims = 1:6, cells = 500, balanced = TRUE, fast = FALSE) + plot_annotation("Top 6 PC heatmaps for the integrated control samples")
```


### Elbow Plot 

Based on the elbow plot, we will proceed using the first 25 PCs as that captures the majority of variation in the data.


```{r,fig.width=10, fig.height=8}
ElbowPlot(ctrl_integrated, ndims = 35) + ggtitle("Elbow plot of integrated control samples")
```

## UMAP

```{r UMAP} 

# Suppressed warning about a change in the default method default method for RunUMAP in Seurat, from using Python’s UMAP (via reticulate) to the R-native UWOT package with the cosine metric.

ctrl_integrated <- RunUMAP(ctrl_integrated, reduction = "pca", dims = 1:25)

```

### Clustering of UMAP

```{r}
ctrl_integrated <- FindNeighbors(ctrl_integrated, dims = 1:25)
ctrl_integrated <- FindClusters(ctrl_integrated, resolution = 0.5, random.seed = 2000)

head(Idents(ctrl_integrated), 5)

save(ctrl_integrated, file = "../data/ctrl_integrated.RData")
```


### Plotting UMAP Clusters

Unannotated UMAP clusters of integrated control samples:

```{r,fig.width=10, fig.height=8}
DimPlot(ctrl_integrated, reduction = "umap") + ggtitle("Unannotated UMAP clusters of integrated control samples")
```

# Annotation

## Seurat Biomarker Analysis

Positive markers (i.e. those that are highly expressed in that cluster) of the clusters were found computationally. 


```{r}

DefaultAssay(ctrl_integrated) <- "RNA"

if(file.exists("../data/ctrl_markers.RData")){
  
  load(file = "../data/ctrl_markers.RData")

  }else{
    
   ctrl_markers <- FindAllMarkers(ctrl_markers, only.pos = T, min.pct = 0.3, logfc.threshold = 0.3)
    
    #This is a time consuming step, so save data: 
    
    save(Controlmarkers, file = "../data/ctrl_markers.RData")

}

DT::datatable(ctrl_markers)
```

## Violin Plots of Biomarker Expression

Violin plots showing known biomarker expression. All except for GP2 (or its alternate name ZAP75) were found. 

Several genes have very low expression in all clusters. This suggests some cell types such as tuft and neuroendocrine cells are rare/not present, although this may reflect their low abundance in vivo. 

Some clusters show high expression of markers of more than one cell type, reflecting transitions between cell types. Goblet cells may differentiate into club and basal cells based on trajectory analysis [@martinsSinglecellDivisionTracing2024].

BC club cells = intermediate cell population from the differentiation of basal cells into club cells [@ravindraSinglecellLongitudinalAnalysis2021].
 

```{r,fig.width=10, fig.height=12}
# Cell-type markers
list_of_cellmarkers <- list(
  Basal = c("KRT5", "KRT14", "TP63", "DAPL1", "NGFR"),
  proliferatingBasal = c("MKI67", "TOP2A", "CDK1"),
  BC_Club = c("KRT4", "KRT13"),
  Club = c("SCGB1A1", "KRT15", "LYPD2"),
  Ciliated = c("FOXJ1", "CCDC153", "CCDC113", "MLF1", "LZTFL1"),
  Neuroendocrine = c("CHGA", "ASCL1", "SOX9", "ITGA2", "ITGB4"), 
  Tuft = c("POU2F3", "AVIL", "GNAT3", "TRPM5"),
  Ionocyte = c("FOXI1", "CFTR", "ASCL3"),
  Goblet = c("MUC5AC", "MUC5B", "GP2", "SPDEF")
)

DefaultAssay(ctrl_integrated) <- "RNA"

for (cell_type in names(list_of_cellmarkers)) {
  markers <- list_of_cellmarkers[[cell_type]]
  
  p <- VlnPlot(
    object = ctrl_integrated,
    features = markers,
    pt.size = 0,
    ncol = 2
  )

  p <- p + plot_annotation(title = paste("Biomarker Violin Plots:", cell_type, sep = " ")) 
  
  print(p)
  num_markers <- ceiling(length(markers) / 2)  
  
    filename <- paste0("../fig/", cell_type, "_Vln_ch1.png")
  
  ggsave(
    filename = filename,
    plot = p,
    width = 8, 
    height = 4 * num_markers
  )
}
```


## Manual Annotation

Manual annotation of the clusters:

* Cluster 0 → Ciliated
* Cluster 1 → Ciliated
* Cluster 2 → Club (also expresses goblet markers)
* Cluster 3 → Basal
* Cluster 4 → Club (also expresses goblet markers)
* Cluster 5 → BC Club
* Cluster 6 → Ciliated
* Cluster 7 → Ciliated (Unknown Subtype)
* Cluster 8 → Proliferating Basal
* Cluster 9 → Proliferating Basal + Ciliated
* Cluster 10 → Ionocyte
* Cluster 11 → Goblet
* Cluster 12 → Ciliated

The unknown cluster may represent a population of dying/stressed cells that persisted after QC filtering, based on high mitochondrial % (see section below).

```{r}
ctrl_integrated_anno <- RenameIdents(ctrl_integrated, 
                              "0" = "Ciliated", 
                              "1" = "Ciliated", 
                              "2" = "Club",
                              "3" = "Basal",
                              "4" = "Club", 
                              "5" = "Basal-Club Transitional", 
                              "6" = "Ciliated" , 
                              "7" = "Ciliated (Stressed)", 
                              "8" = "Proliferating Basal",
                              "9" = "Proliferating + Ciliated",  
                              "10" = "Ionocyte + Tuft", 
                              "11" = "Goblet",
                              "12" = "Ciliated") 

```

### Annotated UMAP Plot


```{r,fig.width=10, fig.height=8}
DefaultAssay(ctrl_integrated_anno) <- "integrated"

annoplot <-  DimPlot(ctrl_integrated_anno, cols = cellColors) + ggtitle("Annotated UMAP of Integrated Control Samples")
annoplot

```



# Additional Plots

## UMAP Grouped by Sample

```{r,fig.width=10, fig.height=8}


p1 <- DimPlot(ctrl_integrated_anno, reduction = "umap", group.by = "sampleType",cols = sampleColors ) + ggtitle("UMAP Coloured by Sample")
p1
```

## UMAP Split by Sample

```{r,fig.width=20, fig.height=8}
p1 <- DimPlot(ctrl_integrated_anno, reduction = "umap", cols = cellColors, split.by = "sampleType") + ggtitle("UMAP Split by Sample")
p1
```


## Mitochondrial Genes Feature Plot

Unknown cluster has high mitochondrial percentage - possibly stressed/dying cells. 

```{r,fig.width=10, fig.height=8}
mitplot <- FeaturePlot(ctrl_integrated_anno, features = "percent.mt") + ggtitle("")
ggsave(
  "../fig/mitplot.pdf",
  plot = mitplot,
  device = NULL,
  path = NULL,
  scale = 1,
  width = 8,
  height = 8,
  units = c("in", "cm", "mm", "px"),
  dpi = 300,
  limitsize = TRUE,
  bg = NULL,
  create.dir = FALSE
)
```

```{r,fig.width=20, fig.height=8}
FeaturePlot(ctrl_integrated_anno, features = "percent.mt", min.cutoff = "q9", split.by = "sampleType") + plot_annotation("Mitochondrial Feature Plot Split by Control Sample") + theme(legend.position = "right")
```

## Violin Plots of Annotated Clusters

Can see overlap between basal, proliferating basal and BC club.
Some overlap between club and goblet cells.


```{r, fig.width=13, fig.height=4}
DefaultAssay(ctrl_integrated_anno) <- "RNA"

short_names <- c(
  "Ciliated" = "CI", 
  "Club" = "CU",
  "Basal" = "BA",
  "Basal-Club Transitional" = "BC", 
  "Ciliated (Stressed)" = "CS",
  "Proliferating Basal" = "PB",
  "Proliferating + Ciliated" = "PC",
  "Ionocyte + Tuft" = "IT",
  "Goblet" = "GO"
)
# Create NEW color vector with short names
new_cellColors <- cellColors
names(new_cellColors) <- short_names[names(cellColors)]

# Verify
print(new_cellColors)

# Permanently rename clusters
ctrl_integrated_anno_short <- RenameIdents(ctrl_integrated_anno, short_names)
library(ggtext)

for (cell_type in names(list_of_cellmarkers)) {
  markers <- list_of_cellmarkers[[cell_type]]

  p <- VlnPlot(
    object = ctrl_integrated_anno_short,
    features = markers,
    pt.size = 0, cols = new_cellColors, ncol = 5
  ) &
  theme(axis.title.y = element_blank(),  # & affects ALL subplots
        axis.text.x = element_text(size = 10))

  p <- p + plot_annotation(title = paste("<b>Title</b>", cell_type, sep = " "))  

  print(p)
}
```

```{r}
library(ggplot2)
library(patchwork)
library(ggtext)

# Create alphabetical prefixes (a), b), c), ...)
letters_prefix <- paste0(letters, ")")

for (i in seq_along(names(list_of_cellmarkers))) {
  cell_type <- names(list_of_cellmarkers)[i]
  markers <- list_of_cellmarkers[[cell_type]]
  
  # Create base plot
  p <- VlnPlot(
    object = ctrl_integrated_anno_short,
    features = markers,
    pt.size = 0,
    cols = new_cellColors,
    ncol = 5
  ) &
    theme(
      axis.title.y = element_blank(),
      axis.text.x = element_blank(),
      plot.margin = margin(t = 20) # Add top margin for annotation
    )
  
  # Add bold alphabetical annotation (larger and lowered)
  p <- p + 
    plot_annotation(
      title = paste0("<b style='font-size:14pt'>", letters_prefix[i], "</b>")
      )
  
  # Print and save
  print(p)
  ggsave(
    filename = paste0("violin_", letters_prefix[i], ".pdf"),
    plot = p,
    width = 15,
    height = 5,
    device = "pdf"
  )
}
```

```{r}
library(ggplot2)
library(patchwork)
library(ggtext)

# Create alphabetical prefixes (a), b), c), ...)
letters_prefix <- paste0(letters, ")")

# Create list to store all plots
plot_list <- list()

for (i in seq_along(names(list_of_cellmarkers))) {
  cell_type <- names(list_of_cellmarkers)[i]
  markers <- list_of_cellmarkers[[cell_type]]
  
  # Create base plot
  p <- VlnPlot(
    object = ctrl_integrated_anno_short,
    features = markers,
    pt.size = 0,
    cols = new_cellColors,
    ncol = 5
  ) &
    theme(
      axis.title.y = element_blank(), # Remove x-axis labels
      axis.title.x = element_blank(), # Remove x-axis title
      plot.title = element_text(size = 14, face = "bold",  # Bold and larger
                               margin = margin(b = 10)),  # Add space below
      plot.margin = margin(t = 20, unit = "pt")  # Add top margin
    )
  
  # Add annotation as plot title (works better than plot_annotation)
  plot_list[[i]] <- p + plot_annotation(letters_prefix[i])
}

# Combine all plots vertically
combined_plot <- wrap_plots(plot_list, ncol = 1)
combined_plot

# Save as single PDF
ggsave(
  filename = "combined_violin_plots.pdf",
  plot = combined_plot,
  width = 15, 
  height = 2.5 * length(plot_list), # Adjust height based on number of plots
  device = "pdf"
)
```


## Cell-type Proportions Barplots

* The following cell-type proportions plot suggest that COPD patients have relatively higher number of goblet cells, which recapitulates the disease in vivo [@shaykhievEmergingBiologyPersistent2019].
* There is marked increase in ciliated cells, which contradicts some literature that suggests that a loss of ciliated epithelial cells is typical in COPD [@rabyMechanismsAirwayEpithelial2023].
* There is an increase in the proportion of the unknown cell population in the COPD samples. If this cell population corresponds to dying cells, this may be an effect of COPD, though it may be due to sample handling and other factors.

### All Cell Types

```{r,fig.width=10, fig.height=8}

ctrl_integrated_anno2 <-ctrl_integrated_anno
DefaultAssay(ctrl_integrated_anno2) <- "RNA"

Idents(ctrl_integrated_anno2) <- factor(Idents(ctrl_integrated_anno2),
                                 levels = c("Ionocyte",
                                             "BC Club", 
                                             "Ciliated",
                                             "Goblet",
                                             "Club",
                                             "Proliferating Basal",
                                             "Proliferating Basal + Ciliated",
                                             "Basal",
                                            "Ciliated (Unknown Subtype)"))


plotCellTypeProps(ctrl_integrated_anno2,clusters = Idents(ctrl_integrated_anno2), sample =  ctrl_integrated_anno2$sampleType) +
  
  scale_fill_manual("Legend", values = cellColors) + ggtitle("All Cell Type Proportions")

```


### Club and Goblet Cells Only

```{r,fig.width=10, fig.height=7}

cell_types <- c("Club", "Goblet")
filtered_cells <- which(Idents(ctrl_integrated_anno2) %in% cell_types)

filtered_clusters <- Idents(ctrl_integrated_anno2)[filtered_cells]
filtered_sample <- ctrl_integrated_anno2$sampleType[filtered_cells]


plotCellTypeProps(ctrl_integrated_anno2, clusters = filtered_clusters, sample = filtered_sample) +
  scale_fill_manual("Legend", values =  cellColors[c("Club", "Goblet")]
) + ggtitle("Goblet and Club Cell Relative Proportions")
```


### Goblet Cells Only

```{r,fig.width=14, fig.height=7}
transformed_props <- getTransformedProps(clusters = Idents(ctrl_integrated_anno2), sample = ctrl_integrated_anno2$sampleType)

props = transformed_props$Proportions

prop_goblet <- as.data.frame(props) %>% filter(clusters == "Goblet")

p <- ggplot(prop_goblet, aes(x = sample, y = Freq, fill = sample)) +
  geom_bar(stat = "identity") +
  geom_text(aes(label = paste0(signif(Freq, 2))), hjust = -0.2, size = 3.5, color = "black") +
  labs(title = "Goblet cell proportions in control samples",
       x = "Sample",
       y = "Proportion of goblet cells") +
  theme_minimal() +
  coord_flip() +
  scale_fill_manual("Sample", values = sampleColors)

p
```

# References

```{r}
all_markers_mine <- FindAllMarkers(ctrl_integrated_anno, min.pct = 0)
saveRDS(all_markers, "../data/all_ctrl_markers_mine.Rds")
```
```{r}
short_names <- c(
  "Ciliated" = "1. Ciliated", 
  "Club" = "2. Club",
  "Basal" = "3. Basal",
  "Basal-Club Transitional" = "4. Basal-Club Transitional", 
  "Ciliated (Stressed)" = "5. Ciliated (Stressed)",
  "Proliferating Basal" = "6. Proliferating Basal",
  "Proliferating + Ciliated" = "7. Proliferating + Ciliated",
  "Ionocyte + Tuft" = "8. Ionocyte + Tuft",
  "Goblet" = "9. Goblet"
)
# Create NEW color vector with short names
new_cellColors <- cellColors
names(new_cellColors) <- short_names[names(cellColors)]

# Verify
print(new_cellColors)

# Permanently rename clusters
#ctrl_integrated_anno <- RenameIdents(ctrl_integrated_anno, short_names)

t1  <-  DimPlot(ctrl_integrated_anno, cols = new_cellColors) 
t1

paper_names <- c("1.HBEC D1 Control" = "HBEC 1 (Control)", "2. HBEC D1 3uM" = "HBEC 1 (3 μM)","3. HBEC D1 10uM" = "HBEC 1 (10 μM)", "4. COPD D1 Control" = "COPD 1 (Control)", "5. COPD D1 3uM" = "COPD 1 (3 μM)", "6. COPD D1 10uM" = "COPD 1 (10 μM)", "7. COPD D2 Control" = "COPD 2 (Control)", "8. COPD D2 3uM" = "COPD 2 (3 uM)", "9. COPD D2 10uM" = "COPD 2(10 uM)")
ctrl_integrated_anno$sampleType <- recode(ctrl_integrated_anno$sampleType,"1. HBEC D1 Control" = "HBEC 1 (Control)", "2. HBEC D1 3uM" = "HBEC 1 (3 μM)","3. HBEC D1 10uM" = "HBEC 1 (10 μM)", "4. COPD D1 Control" = "COPD 1 (Control)", "5. COPD D1 3uM" = "COPD 1 (3 μM)", "6. COPD D1 10uM" = "COPD 1 (10 μM)", "7. COPD D2 Control" = "COPD 2 (Control)", "8. COPD D2 3uM" = "COPD 2 (3 uM)", "9. COPD D2 10uM" = "COPD 2(10 uM)")
t2<- DimPlot(ctrl_integrated_anno, split.by = "sampleType", ncol = 1, cols = cellColors) +NoLegend() 
t2
```

```{r}
# Get cells for the specific cluster
stressed_cells <- subset(ctrl_integrated_anno, idents = "5. Ciliated (Stressed)")

# Calculate statistics
mean_mt <- mean(stressed_cells$percent.mt, na.rm = TRUE)
median_mt <- median(stressed_cells$percent.mt, na.rm = TRUE)

# Print formatted results
cat(sprintf(
  "Ciliated (Stressed) QC metrics:
  Mean mitochondrial percentage = %.2f%%
  Median mitochondrial percentage = %.2f%%",
  mean_mt, median_mt
))
```


```{r, include=FALSE}
ggsave(
  "../fig/ctrl_mine.pdf",
  plot = t1,
  device = NULL,
  path = NULL,
  scale = 1,
  width = 22,
  height = 16,
  units = "cm",
  dpi = 300,
  limitsize = TRUE,
  bg = NULL,
  create.dir = FALSE
)


ggsave(
  "../fig/ctrl_mine_split.pdf",
  plot = t2,
  device = NULL,
  path = NULL,
  scale = 1,
  width = 18,
  height = 45,
  units = "cm",
  dpi = 600,
  limitsize = TRUE,
  bg = NULL,
  create.dir = FALSE
)

# saveRDS(ctrl_integrated_anno, file = "/data/jx303/sysbioproject/data/ctrl_integrated_anno.rds")
```

